src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let stateAtTheMoment = { ...state }; // clone the state at the beginning
+  const stateHistory = []; // const, because an array is an object in JS
+
+  for (const action of actions) {
+    let newState = { ...stateAtTheMoment };
+    // Clone the current state to the new state
+
+    if (action.type === 'clear') {
+      newState = {};
+    }
+
+    if (action.type === 'addProperties') {
+      Object.assign(newState, action.extraData);
+    }
+
+    if (action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete newState[key];
+      }
+    }
+
+    stateHistory.push(newState);
+    // push the new state to the empty array, where the result is collected
+
+    stateAtTheMoment = newState; // update the current state to the new state
```

Comment:
You should avoid reassigning `stateAtTheMoment` with `newState` at the end of each loop iteration. Instead, push a copy of the object to the `stateHistory` array at the end of each loop cycle, but outside of the `switch` block, after the current action is processed.